binary_search_tree/tree.py

```diff
@@ -9,50 +13,144 @@ def __init__(self, key, val = None):
+        return current_node
+            
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+            
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: 0(log n)
+    # Space Complexity: 0(log n)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key <key:
+                current = current.right
+            else:
+                current = current.left
+        return None
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(n)
+    def inorder_helper(self, current_node, items):
+        if current_node != None:
+            self.inorder_helper(current_node.left, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.inorder_helper(current_node.right, items)
+        
     def inorder(self):
-        pass
+        #list of items to be returned --> depth-first traversal sorted key values in an ascending order, Left,Current,Right
+        items = []
+        self.inorder_helper(self.root, items)
+        return items
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(n)   
+    def preorder_helper(self, current_node, items):
+        if current_node == None:
+            return items
+        else:
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.preorder_helper(current_node.left, items)
+            self.preorder_helper(current_node.right, items)
+            return items
+     
     def preorder(self):
-        pass
+        #list of items to be returned --> depth-first traversal sorted current, left, right
+        items = []
+        if self.root:
+            self.preorder_helper(self.root, items)
+        return items
+
+
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(n)  
+    def postorder_helper(self, current_node, items): 
+        if current_node:
+            self.postorder_helper(current_node.left, items)
+            self.postorder_helper(current_node.right, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+        return items
+      
+    def postorder(self): 
+        #list of items to be returned --> depth-first traversal sorted left, right, current
+        items = []
+        if self.root:
+            self.postorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def postorder(self):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: 0(n)
+    # Space Complexity: 0(n)  
```

Comment:
This is true assuming an unbalanced tree, but if we assume a balanced tree, then we only need to recurse down O(log(n)) times along each branch of the tree, so we can use that for the time and space complexity calculations.

Reply:
When I gave this comment, I was working off the instructor solution, but I now think this is wrong. The time complexity is O(n) regardless because we have to travel to every node, and O(1) regardless because no additional space is allocated.

Sorry for any confusion.